### PR TITLE
Start collecting usage data on features

### PR DESCRIPTION
**Pull request by rli**
## Types of changes

- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description
Changes:
* Improve MetricEvent builder ergonomics:
    * events can't be built without supplying a name(space)
    * event names are trimmed of illegal characters
    * a datum will be generated for a namespace during flush if none are supplied
* Wrapper classes:
    * AnActionWrapper
        * records action performed
    * ComboBoxActionWrapper
        * records action performed
    * ToggleActionWrapper
        * records action performed
    * LoggingDialogWrapper
        * records OK/Cancel action
    * These cover most, if not all, of the actions currently available to a user.
    * Names are generated based on the implementation class if `getNamespace()` is not overridden
    * relevant abstract methods are marked final so implementation classes cannot accidentally stop recording events
        * do\<original-name\>() is the new entry point
* Additional events:
    * SamInvoke: was run or debug? did the user hit a packaging error? what was the runtime?
    * SamDeploy: for SAM CLI {build, package, deploy}, was the command successful?
    * SamProjectInit: what was the template? runtime?

Actual errors are not collected if something failed, since they are potentially sensitive.

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/sam/SamInvokeRunner.kt [hunk 1]
```diff
@@ -92,6 +94,17 @@ class SamInvokeRunner : AsyncProgramRunner<RunnerSettings>() {
...
+            }.whenComplete { _, exception ->
+                telemetry.record("SamInvoke") {
+                    val type = if (environment.isDebug()) "Debug" else "Run"
+                    datum(type) {
+                        value(1.0)
+                        unit(MetricUnit.COUNT)
+                        // exception can be null but is not annotated as nullable
+                        metadata("hasException", (exception != null).toString())
+                        metadata("runtime", state.settings.runtime.name)
+                    }
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/sam/SamInvokeRunner.kt]: Add in if it was template based or not

File: core/src/software/aws/toolkits/core/telemetry/MetricEvent.kt [hunk 2]
```diff
@@ -37,18 +38,25 @@ interface MetricEvent {
 
             fun metadata(key: String, value: String): Builder
 
+            fun metadata(key: String, value: Boolean?): Builder = metadata(key, value.toString())
```
Comment (abrooksv) [inline core/src/software/aws/toolkits/core/telemetry/MetricEvent.kt]: Where do we use null for the boolean?

File: core/src/software/aws/toolkits/core/telemetry/TelemetryNamespace.kt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.core.telemetry
+
+interface TelemetryNamespace {
+    fun getNamespace(): String = javaClass.name.split(".").last()
```
Comment (abrooksv) [inline core/src/software/aws/toolkits/core/telemetry/TelemetryNamespace.kt]: Is this not the same ass javaClass.simpleName?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                value(1.0)
+                unit(MetricUnit.COUNT)
+            }
+        }
+    }
+
+    open fun doActionPerformed(e: AnActionEvent) = super.actionPerformed(e)
+
+    companion object {
+        val telemetry = TelemetryService.getInstance()
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt]: protected

File: jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                value(1.0)
+                unit(MetricUnit.COUNT)
+            }
+        }
+    }
+
+    abstract fun doActionPerformed(e: AnActionEvent)
+
+    companion object {
+        val telemetry = TelemetryService.getInstance()
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt]: protected

File: jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            }
+        }
+    }
+
+    abstract fun doIsSelected(e: AnActionEvent): Boolean
+
+    abstract fun doSetSelected(e: AnActionEvent, state: Boolean)
+
+    companion object {
+        val telemetry = TelemetryService.getInstance()
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt]: protected

File: jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/LoggingDialogWrapper.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        telemetry.record(getNamespace()) {
+            datum("CancelAction") {
+                value(1.0)
+                unit(MetricUnit.COUNT)
+            }
+        }
+    }
+
+    companion object {
+        val telemetry = TelemetryService.getInstance()
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/LoggingDialogWrapper.kt]: protected

File: jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/LoggingDialogWrapper.kt [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        super(project, canBeParent, applicationModelIfPossible)
+
+    constructor(parent: Component, canBeParent: Boolean):
+        super(parent, canBeParent)
+
+    override fun doOKAction() {
+        super.doOKAction()
+
+        telemetry.record(getNamespace()) {
+            datum("OKAction") {
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/LoggingDialogWrapper.kt]: We should have a hook here for collecting data from the dialog for things like CreateLambda runtime

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/EditFunctionDialog.kt [hunk 9]
```diff
@@ -335,6 +335,8 @@ class EditFunctionDialog(
         }
     }
 
+    override fun getNamespace(): String = "EditFunctionDialog"
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/EditFunctionDialog.kt]: We should differentiate Create vs Edit.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/wizard/SamInitProjectBuilderCommon.kt [hunk 10]
```diff
@@ -47,12 +48,25 @@ abstract class SamProjectTemplate {
...
+
+    protected open fun doBuild(runtime: Runtime, outputDir: VirtualFile) {
         SamInitRunner(SamModuleType.ID, outputDir, runtime).execute()
     }
 
     fun getModuleBuilderProjectTemplate(builder: ModuleBuilder) =
             SamProjectTemplateWrapper(this, builder)
+
+    companion object {
+        val telemetry = TelemetryService.getInstance()
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/ui/wizard/SamInitProjectBuilderCommon.kt]: should be private

File: jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/LoggingDialogWrapper.kt (hunk 8 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/LoggingDialogWrapper.kt]: Good idea, will follow up in another commit.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/EditFunctionDialog.kt (hunk 9 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/EditFunctionDialog.kt]: fixed

File: jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+    final override fun setSelected(e: AnActionEvent, state: Boolean) {
+        doSetSelected(e, state)
+        telemetry.record(getNamespace()) {
+            datum(e.place) {
+                value(1.0)
+                unit(MetricUnit.COUNT)
+            }
+        }
+    }
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt]: Can you wrap the report of count metrics to an extension method? There is no point to repeat three times for the same thing.

Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/ActionWrappers.kt]: updated

File: jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/LoggingDialogWrapper.kt [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+//  @Deprecated
+//  protected DialogWrapper(boolean canBeParent, boolean applicationModalIfPossible)
+//    this(null, canBeParent, applicationModalIfPossible);
+//  }
+//  protected DialogWrapper(Project project, boolean canBeParent, boolean applicationModalIfPossible) {
+//    <logic>
+//  }
+//  protected DialogWrapper(@NotNull Component parent, boolean canBeParent) {
+//    <logic>
+//  }
```
Comment (zhangzhx) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/LoggingDialogWrapper.kt]: Do we need all these comments? Doesn't really seem very useful to me.

Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/LoggingDialogWrapper.kt]: I don't mind removing them. It's just easier to tell what the original constructors are when you don't have to try to find them in the parent's source code

Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/components/telemetry/LoggingDialogWrapper.kt]: The constructors are just a set of different forms for supporting different parameter signatures. It is pretty strait forward I think. I don't mind leaving it there, it's just doesn't standard docs :)

**Pull request merged**